ngraph/core/reference/include/ngraph/runtime/reference/convolution_backprop_data.hpp

```diff
@@ -0,0 +1,309 @@
+                        row_major_strides(filter_shape).at(filter_in_channel_axis);
+
+                    while (in_it != in_it_end && filter_it != filter_it_end)
+                    {
+                        const Coordinate& in_coord = *in_it;
+                        if (in_transform.has_source_coordinate(in_coord))
+                        {
+                            size_t in_idx = in_transform.index(in_coord);
+                            const Coordinate& filter_coord = *filter_it;
+                            size_t filter_idx = filter_transform.index(filter_coord);
+                            for (size_t in_channel = 0; in_channel < n_in_channels; ++in_channel)
+                            {
+                                ACCUMULATION in_v = static_cast<ACCUMULATION>(in[in_idx]);
+                                ACCUMULATION f_v = static_cast<ACCUMULATION>(filter[filter_idx]);
+
+                                result += in_v * f_v;
+                                in_idx += in_channel_stride;
+                                filter_idx += filter_in_channel_stride;
+                            }
+                        }
+                        ++in_it;
+                        ++filter_it;
+                    }
+
+                    out[out_transform.index(out_coord)] = result;
+                }
+                std::fesetround(old_mode);
+            }
+
+            template <typename OUTPUT,
+                      typename FILTER,
+                      typename INPUT,
+                      typename ACCUMULATION = typename widen<INPUT>::type>
+            void convolution_backprop_in(const OUTPUT* delta_out,
+                                         const FILTER* filter,
+                                         INPUT* delta_in,
+                                         const Shape& out_shape,
+                                         const Shape& filter_shape,
+                                         const Shape& in_shape,
+                                         const Strides& in_dilation,
+                                         const Strides& filter_dilation,
+                                         const CoordinateDiff& forward_in_pad_bellow,
+                                         const CoordinateDiff& forward_in_pad_above,
+                                         const Strides& stride)
+            {
+                // Note that we only reverse the spatial dimensions here (loop
+                // starts at 2)
+                std::vector<INPUT> reversed(shape_size(filter_shape));
+                AxisSet reverse_axes;
+                size_t reverse_axes_start = 2;
+                for (size_t i = reverse_axes_start; i < filter_shape.size(); ++i)
+                {
+                    reverse_axes.insert(i);
+                }
+                reverse(reinterpret_cast<const char*>(filter),
+                        reinterpret_cast<char*>(&reversed[0]),
```

Comment:
ConvolutionBackpropData will be addressed in different ticket, so do not review it's code - I didn't change it only moved it to other file because previously it was together with convolution.